staging/src/k8s.io/kubectl/pkg/cmd/create/create_service.go

```diff
@@ -45,6 +54,156 @@ func NewCmdCreateService(f cmdutil.Factory, ioStreams genericclioptions.IOStream
 	return cmd
 }
 
+// ServiceOpts holds the options for 'create service' sub command
+type ServiceOpts struct {
+	PrintFlags *genericclioptions.PrintFlags
+	PrintObj   func(obj runtime.Object) error
+
+	Name         string
+	TCP          []string
+	Type         v1.ServiceType
+	ClusterIP    string
+	NodePort     int
+	ExternalName string
+
+	FieldManager     string
+	CreateAnnotation bool
+	Namespace        string
+
+	Client         clientCoreV1.CoreV1Interface
+	DryRunStrategy cmdutil.DryRunStrategy
+	DryRunVerifier *resource.DryRunVerifier
+	genericclioptions.IOStreams
+}
+
+// Complete completes all the required options
+func (o *ServiceOpts) Complete(f cmdutil.Factory, cmd *cobra.Command, args []string) error {
+	name, err := NameFromCommandArgs(cmd, args)
+	if err != nil {
+		return err
+	}
+	o.Name = name
+
+	cs, err := f.KubernetesClientSet()
+	if err != nil {
+		return err
+	}
+	o.Client = cs.CoreV1()
+
+	o.Namespace, _, err = f.ToRawKubeConfigLoader().Namespace()
+	if err != nil {
+		return err
+	}
+
+	o.DryRunStrategy, err = cmdutil.GetDryRunStrategy(cmd)
+	if err != nil {
+		return err
+	}
+	dynamicClient, err := f.DynamicClient()
+	if err != nil {
+		return err
+	}
+	discoveryClient, err := f.ToDiscoveryClient()
+	if err != nil {
+		return err
+	}
+	o.DryRunVerifier = resource.NewDryRunVerifier(dynamicClient, discoveryClient)
+	cmdutil.PrintFlagsWithDryRunStrategy(o.PrintFlags, o.DryRunStrategy)
+
+	printer, err := o.PrintFlags.ToPrinter()
+	if err != nil {
+		return err
+	}
+
+	o.PrintObj = func(obj runtime.Object) error {
+		return printer.PrintObj(obj, o.Out)
+	}
+
+	return nil
+}
+
+func (o *ServiceOpts) createService() (*v1.Service, error) {
+	if o.ClusterIP == v1.ClusterIPNone && o.Type != v1.ServiceTypeClusterIP {
+		return nil, fmt.Errorf("ClusterIP=None can only be used with ClusterIP service type")
+	}
+	if o.ClusterIP != v1.ClusterIPNone && len(o.TCP) == 0 && o.Type != v1.ServiceTypeExternalName {
+		return nil, fmt.Errorf("at least one tcp port specifier must be provided")
+	}
+	if o.Type == v1.ServiceTypeExternalName {
+		if errs := validation.IsDNS1123Subdomain(o.ExternalName); len(errs) != 0 {
+			return nil, fmt.Errorf("invalid service external name %s", o.ExternalName)
+		}
+	}
+
+	ports := []v1.ServicePort{}
+	for _, tcpString := range o.TCP {
+		port, targetPort, err := generateversioned.ParsePorts(tcpString)
```

Comment:
We're trying not to rely on generate package, and you've just exposed a method from that package.

Reply:
Yes, that is true. I didn't want to duplicate the code.
Should I duplicate the method into create_service.go instead?
And make ParsePorts in the generator private again?

Reply:
yes, please. the generate package will be remove once all of those changes land.